DM-33278: Improve storage class conversion

timj opened the pull request:
* Add a dataset type compatibility method.
* Add support for specifying methods.

## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

timj committed:
Support methods in storage class converters

If a converter ends with () it is assumed to be a method
to run on the object to be converted and not a class method
or function.

timj committed:
Convert python type on put in in-memory datastore

This removes surprises when someone gets the dataset (there
is no need to also add conversion on get), and provides
consistent application of parameters.

timj committed:
Add DatasetType.is_compatible_with method

timj committed:
Add a converter for dict -> Packages

timj committed:
Add news fragment

timj committed:
Add some short docs on storage class conversion.

timj committed:
Include DatasetType equality test involving isCalibration

andy-slac commented on doc/lsst.daf.butler/formatters.rst:
```diff
@@ -139,6 +146,22 @@ You can do this using YAML anchors and references but the preferred approach is
+        lsst.daf.base.PropertySet: lsst.pipe.base.TaskMetadata.from_metadata
+
+In the first definition, the configuration says that if a ``PropertySet`` object is given then the ``toDict()`` method can be called on it and the returned value will be a `dict`.
+In the second definition a ``PropertySet`` is again specified but this time the ``from_metadata`` class method will be called with the ``PropertySet`` as the first parameter and a ``TaskMetadata`` will be returned.
```
This difference between `()` and no `()` confused me a lot when I first saw it and it was not obvious what it means. I had to read everything more than once to figure it out. Now I think that `toDict()` is basically the same as `lsst.daf.base.PropertySet.toDict` (non-bound method). Would it be better to drop `()` and use non-bound method syntax for symmetry?

timj replied:
I've been pondering this and the `()` syntax has the advantage that it doesn't lead to duplicating the full class name on both sides:

```yaml
lsst.daf.butler.tests.MetricsExample: exportAsDict()
```
vs
```yaml
lsst.daf.butler.tests.MetricsExample: lsst.daf.butler.tests.MetricsExample.exportAsDict
```
The latter does work and doesn't require any code changes to make it work without this ticket. Maybe that's a good enough argument for not trying to bother with the more succinct form.

andy-slac commented on python/lsst/daf/butler/core/datasets/type.py:
```diff
@@ -240,25 +240,78 @@ def __repr__(self) -> str:
-        if self._isCalibration != other._isCalibration:
+        return True
+
+    def is_compatible_with(self, other: "DatasetType") -> bool:
```
Do you need to quote `DatasetType` here?

timj replied:
I think I'm confused between `from __future__ import annotations` being enabled and not being enabled. There was some other place where we did have to use strings if the type was used in the context of its own class but maybe that was to do with pydantic.

TallJimbo replied:
I believe we cannot use `from __future__ import annotations` with pydantic (which is really a shame, and I think it means we should mostly put pydantic classes in their own modules, so we can use it everywhere else).  And without `__future__.annotations`, you do have to quote in places like this.

andy-slac commented on python/lsst/daf/butler/core/datasets/type.py:
```diff
@@ -240,25 +240,78 @@ def __repr__(self) -> str:
-        if self._isCalibration != other._isCalibration:
+        return True
+
+    def is_compatible_with(self, other: "DatasetType") -> bool:
```
Does "compatible" mean compatible for both `get` and `put`? Is it symmetrical (X.is_compatible_with(Y) === Y.is_compatible_with(X))?

timj replied:
It really means `self_sc.can_convert(other_sc)` -- You have made me realize that ctrl_mpexec is going to have to try to distinguish inputs (so can convert on get to the required type for the Task) and outputs (so can convert the type used in the Task to the required type in the butler).

andy-slac reviewed (approved):
Looks good, couple of minor comments/questions.

timj committed:
Improve conversion docs

timj committed:
Revert "Support methods in storage class converters"

This reverts commit 8e15cb70fafab1f9a54ec1c809d281a0bec3fa95.

We can do the same thing with a full name of an unbound method.

timj committed:
Add explicit test of unbound method converter

timj committed:
Add experimental warning to storage class conversion docs

timj committed:
Update the documentation to indicate that unbound methods are allowed

timj committed:
Allow StorageClass.can_convert to return true if pytypes match

Sometimes a storage class can be defined with a different
name but is referring to the same python type.
